File: Problems/Strings/Group-Anagrams/Group-Anagrams.cpp
#include <iostream>
#include <vector>
#include <string>
#include <unordered_map>
#include <algorithm>

// Using hash map - Time complexity: O(NKlogK), where N is the length of strs, and K is the maximum length of a string in strs.

class Solution {
public:
    std::vector<std::vector<std::string>> groupAnagrams(std::vector<std::string> &strs)
    {
        std::unordered_map<std::string, int> m;
        std::vector<std::vector<std::string>> ans;

        for (auto &s : strs)
        {
            auto key = s;
            sort(begin(key), end(key));
            if (!m.count(key))
            {
                m[key] = ans.size();
                ans.emplace_back();
            }
            ans[m[key]].push_back(s);
        }

        std::reverse(ans.begin(), ans.end());
        return ans;
    }
};

int main()
{
    std::vector<std::string> input = {"eat", "tea", "tan", "ate", "nat", "bat"};
    std::cout << "Input: [";
    for (const std::string &word : input)
    {
        std::cout << word << "";
        if (word != input.back())
        {
            std::cout << ", ";
        }
    }
    std::cout << "]" << std::endl;


    Solution sol;
    std::vector<std::vector<std::string>> result = sol.groupAnagrams(input);

    std::cout << "Output: [";

    for (const std::vector<std::string> &group : result)
    {
        std::cout << "[";
        for (const std::string &word : group)
        {
            std::cout << word << "";
            if (word != group.back())
            {
                std::cout << ", ";
            }
        }
        std::cout << "]";
        if (group != result.back())
        {
            std::cout << ", ";
        }
    }
    std::cout << "]" << std::endl;

    return 0;
}

File: Problems/Strings/Group-Anagrams/Group-Anagrams.cs
using System;
using System.Linq;
using System.Collections.Generic;

// Using hash map - Time complexity: O(NKlogK), where N is the length of strs, and K is the maximum length of a string in strs.
public class Solution
{
    public IList<IList<string>> GroupAnagrams(string[] strs)
    {
        Dictionary<string, int> m = new Dictionary<string, int>();
        List<IList<string>> ans = new List<IList<string>>();

        for (int i = 0; i < strs.Length; i++)
        {
            string s = strs[i];
            string key = new string(s.ToCharArray());
            char[] keyChars = key.ToCharArray();
            Array.Sort(keyChars);
            key = new string(keyChars);

            if (!m.ContainsKey(key))
            {
                m[key] = ans.Count;
                ans.Add(new List<string>());
            }

            ans[m[key]].Add(s);
        }

        ans.Reverse();
        return ans;
    }
}

class Program
{
    static void Main(string[] args)
    {
        
        // Ejemplo de uso:
        string[] input = { "eat", "tea", "tan", "ate", "nat", "bat" };
        Console.Write("Input: strs = ["); 
        Console.Write(string.Join(", ", input));
        Console.WriteLine("]");

        Solution sol = new Solution();
        IList<IList<string>> result = sol.GroupAnagrams(input);

        Console.Write("Output: [");
        foreach (var group in result)
        {
            Console.Write("[" + string.Join(", ", group) + "]");
            if (group != result.Last())
            {
                Console.Write(", ");
            }
        }
        Console.WriteLine("]");
    }
}